internal/dbusmgr/dbusmgr.go

```diff
@@ -0,0 +1,111 @@
+// +build linux
+// Code in this package is heavily adapted from https://github.com/opencontainers/runc/blob/7362fa2d282feffb9b19911150e01e390a23899d/libcontainer/cgroups/systemd
+// Credit goes to the runc authors.
+
+package dbusmgr
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync"
+
+	systemdDbus "github.com/coreos/go-systemd/v22/dbus"
+	dbus "github.com/godbus/dbus/v5"
+)
+
+var (
+	dbusC        *systemdDbus.Conn
+	dbusMu       sync.RWMutex
+	dbusInited   bool
+	dbusRootless bool
+)
+
+type DbusConnManager struct {
+}
+
+// NewDbusConnManager initializes systemd dbus connection manager.
+func NewDbusConnManager(rootless bool) *DbusConnManager {
+	if dbusInited && rootless != dbusRootless {
+		panic("can't have both root and rootless dbus")
```

Comment:
Nice, I was thinking about adding something like this but never got to it, mostly because it's not possible in a short-lived binary like `runc` (which is either rootless or not, can't be both).

Reply:
But you do need to
(1) take a mutex before getting to these variables, otherwise races are possible (if two goroutines call `NewDbusConnManager`)
(2) do set `dbusInited = true` -- otherwise this check never works.

Reply:
hahaha then you can thank your past-self https://github.com/opencontainers/runc/blame/2f8e8e9d9775003bf76730a8a1c10f29b0000edf/libcontainer/cgroups/systemd/dbus.go#L25

Reply:
🤦🏻

Reply:
https://github.com/opencontainers/runc/pull/3006